stix_shifter/stix_transmission/src/modules/bigfix/bigfix_status_connector.py

```diff
@@ -51,39 +49,51 @@ def create_status_connection(self, search_id):
 
                     reporting_agents = int(response_dict.get('reportingAgents', '0'))
                     total_results = int(response_dict.get('totalResults', '0'))
-
-                    if self.DEFAULT_CLIENT_COUNT == client_count:
-                        return_obj['progress'] = 0
-                    else:
-                        progress = (reporting_agents / client_count) * 100
-                        progress_floor = math.floor(progress)
-                        return_obj['progress'] = progress_floor
-
+                    return_obj = self._get_progress_status(client_count, reporting_agents, return_obj)
                     if client_count <= reporting_agents:
                         return_obj['status'] = Status.COMPLETED.value
                         if total_results <= 0:
                             return_obj['status'] = Status.ERROR.value
-
-                    if return_obj['success'] == True and return_obj['status'] == Status.COMPLETED.value:
-                        if not self.get_succes_status(response_dict):
-                            return_obj['status'] = Status.ERROR.value
-                            ErrorResponder.fill_error(return_obj, response_dict, ['results', '+isFailure=True','~result'])
-
+                        break
+                    time_iter -= WAIT_PERIOD
+                    if time_iter < WAIT_PERIOD:
+                        time.sleep(time_iter)
+                    else:
+                        time.sleep(WAIT_PERIOD)
                 except json.decoder.JSONDecodeError:
                     response_dict = xmltodict.parse(response_txt)
-                    ErrorResponder.fill_error(return_obj, response_dict, ['BESAPI','ClientQueryResults','QueryResult', '+IsFailure=1','~Result'])
+                    ErrorResponder.fill_error(return_obj, response_dict, ['BESAPI', 'ClientQueryResults',
+                                                                          'QueryResult', '+IsFailure=1', '~Result'])
+                    break
             else:
                 if ErrorResponder.is_plain_string(response_txt):
                     ErrorResponder.fill_error(return_obj, message=response_txt)
+                    break
                 else:
                     raise UnexpectedResponseException
-        
+        return return_obj
+
+    def create_status_connection(self, search_id):
+        response_txt = None
+        return_obj = dict()
+        try:
+            response = self.api_client.get_sync_query_results(self.RELEVANCE)
+            response_txt = response.read().decode('utf-8')
+            client_count = self.DEFAULT_CLIENT_COUNT
+            search = re.search(self.PATTERN, response_txt, re.IGNORECASE)
+            if search:
+                client_count = search.group(1)
+            client_count = int(client_count)
+            return_obj = self.status_api_response(search_id, client_count)
```

Comment:
if I'm not wrong this looks like its not gonna return anything to the UDS until the timeout limit reaches for all the RUNNING status cases. we need to return the status whenever UDS makes a status call even if its running. UDS itself make calls to check the status until the timeout.

Reply:
Hi Azam,

- The DS(data source) API does not provide a timeout parameter with the API calls.  The DS(data source) does not provide status parameter in Asynchronous API calls. There is no persistence of data in connector.
- Hence had implemented an internal wait logic in the status call in the form of a while loop which will accumulate data from data source and check progress against internal timeout(1 minute) with wait time of 6 seconds per loop (or seconds to timeout) whichever is smaller.
- This will immediately send status back if progress reaches 100% before timeout is reached.
- If progress failed to reach 100% it will change the status to completed and send the status back with progress equal to last observed progress.

Reply:
We are clear about the API limitations on timeout. The main concern is, we can't keep the UDS caller waiting to send back the status if progress is less than 100%. That may result in cloud functions timeout. We need to find a way to persist the internal wait logic/count and return the status back to UDS at the same time. That's why I suggested to use the timestamp instead in my other comment.

Reply:
As connectors are invoked via cloud functions we may not be able to persist data on the connector side.

Reply:
Right. You wont be able to persist the timeout in connector side. But, Its the connector's responsibility to return the status back to UDS on time. This is very specific case for BigFix so it needs to be handled in connector side.

Reply:
HI Azam,
As discussed and shared over email we have incorporated the query status completion logic with the latest commit.

Reply:
looks good. thanks!